### PR TITLE
Withdrawals: containers and  process BLS to ETH1 changes

### PR DESCRIPTION
**Pull request by potuz**
This PR is the first in a WIP of implementation of withdrawals, it

- Adds the probuf files needed for the new containers towards withrawals. It adds ssz objects for `Withdrawals`, `BLSToExecutionChange` and `SignedBLSToExecutionChange` messages. 
- Adds a core pure block processing function `ProcessBLSToExecutionChange` that modifies the validator in the beaconstate on site (this function will be called during block processing)
- It adds new Domain  and prefix constants.

### CODE REVIEW

**Comment by terencechain** [comment]
@potuz do you want CI/CD to run with this PR? I'll change it to `WIP:` for now

**Comment by potuz** [comment]
> @potuz do you want CI/CD to run with this PR? I'll change it to `WIP:` for now

It was wip

**Comment by terencechain** [comment]
> > @potuz do you want CI/CD to run with this PR? I'll change it to `WIP:` for now
> 
> It was wip

I thought `[WIP]` or `WIP:` would trigger different behaviors but they all are the same (i.e. not setting the PR to draft correctly).  I wonder if that component is broken. In any case, I had set it to draft unless that's not what you intended

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — DISMISSED** [review]

**Comment by potuz** [comment]
@terencechain thanks for 73df61b...

File: beacon-chain/core/blocks/withdrawals.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	epoch := slots.ToEpoch(st.Slot())
+	domain, err := signing.Domain(st.Fork(), epoch, params.BeaconConfig().DomainBLSToExecutionChange, st.GenesisValidatorsRoot())
+	if err != nil {
+		return nil, err
+	}
+	if err := signing.VerifySigningRoot(message, fromPubkey, signed.Signature, domain); err != nil {
+		return nil, signing.ErrSigFailedToVerify
+	}
+	newCredentials := make([]byte, 12)
```
Comment (rkapka) [inline beacon-chain/core/blocks/withdrawals.go:72]: Let's create a const for the magic number `12`

File: beacon-chain/core/blocks/error.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package blocks
+
+import "github.com/pkg/errors"
+
+var errNilSignedWithdrawalMessage = errors.New("nil SignedBLSToExecutionChange message")
+var errNilWithdrawalMessage = errors.New("nil BLSToExecutionChange message")
+var errInvalidBLSPrefix = errors.New("invalid BLS withdrawal credential prefix")
```
Comment (rkapka) [inline beacon-chain/core/blocks/error.go:7]: How about `withdrawal credential prefix is not a BLS prefix`?

File: beacon-chain/core/blocks/withdrawals.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	val, err := st.ValidatorAtIndex(message.ValidatorIndex)
+	if err != nil {
+		return nil, err
+	}
+	cred := val.WithdrawalCredentials
+	if cred[0] != params.BeaconConfig().BLSWithdrawalPrefixByte {
+		return nil, errInvalidBLSPrefix
+	}
+
+	// hash the public key and verify it matches the withdrawal credentials
```
Comment (rkapka) [inline beacon-chain/core/blocks/withdrawals.go:55]: Can we add a test case for a failure here?

**Review by rkapka — DISMISSED** [review]

**Review by rkapka — APPROVED** [review]

**Pull request merged**